File: java/10_CreateDataset.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17+
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.2

// Recipe 10: create a Dataset and read its rows back.
//
// Demonstrates the v2 Dataset endpoints plus the Table read needed to
// walk the rows:
//   POST /v2/{entity}/{project}/datasets
//       -> create the Dataset, returns (object_id, digest, version_index)
//   GET  /v2/{entity}/{project}/datasets/{object_id}/versions/{digest}
//       -> read Dataset metadata, including a *reference* to its rows
//   POST /table/query
//       -> read the actual rows out of the referenced Table
//
// Wire-level points worth knowing:
//
//   - These are v2 endpoints under /v2/{entity}/{project}/datasets, not a
//     v1-style /datasets/create. Entity + project live in the URL path.
//     Read uses GET (the rest of the API is POST-only); create uses POST.
//   - A Dataset is addressed by (object_id, digest) and is content-
//     addressed — identical (name, rows) collapses to the same version.
//     The name is stamped with a per-run Unix timestamp so every run
//     exercises the write path rather than resolving to an existing object.
//   - The read response's `rows` field is a *reference string* to the
//     underlying Table, not the row data. Parse the table digest out of
//     it and call /table/query. Rows come back wrapped as {digest, val,
//     original_index?} — the actual content lives under `val`.
//
// Run:
//   jbang java/10_CreateDataset.java

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class CreateDataset {
    static final ObjectMapper MAPPER = new ObjectMapper();
    static final HttpClient HTTP = HttpClient.newHttpClient();
    static final String BASE_URL = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai");
    static final Pattern TABLE_REF = Pattern.compile("/table/([A-Za-z0-9_-]+)$");
    static String entity;
    static String project;
    static String projectId;
    static String authHeader;

    static String getenv(String key, String def) {
        String v = System.getenv(key);
        return (v == null || v.isEmpty()) ? def : v;
    }

    static void fail(String msg) {
        System.err.println("FAIL: " + msg);
        System.exit(1);
    }

    // Issues a request with auth, optionally with a JSON body, and decodes the
    // JSON response. Any non-2xx is fatal. post/get wrap it.
    static JsonNode doReq(String method, String path, Object body) throws Exception {
        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(BASE_URL + path))
                .header("Authorization", authHeader);
        if (body != null) {
            b.header("Content-Type", "application/json")
                    .method(method, HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body), StandardCharsets.UTF_8));
        } else {
            b.method(method, HttpRequest.BodyPublishers.noBody());
        }
        HttpResponse<String> res = HTTP.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2)
            fail("HTTP " + res.statusCode() + " for " + method + " " + path + ": " + res.body());
        String body2 = res.body();
        return (body2 == null || body2.isEmpty()) ? MAPPER.createObjectNode() : MAPPER.readTree(body2);
    }

    static JsonNode post(String path, Object body) throws Exception { return doReq("POST", path, body); }
    static JsonNode get(String path) throws Exception { return doReq("GET", path, null); }

    // A dataset row with insertion-ordered keys (question, then answer), so
    // Jackson serializes it the same way as the other language ports.
    static Map<String, Object> qaRow(String question, String answer) {
        var row = new java.util.LinkedHashMap<String, Object>();
        row.put("question", question);
        row.put("answer", answer);
        return row;
    }

    public static void main(String[] args) throws Exception {
        var missing = new ArrayList<String>();
        for (String k : List.of("WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"))
            if (getenv(k, "").isEmpty()) missing.add(k);
        if (!missing.isEmpty())
            fail("Missing required env vars: " + String.join(", ", missing) + ". See ../README.md#setup.");

        entity = System.getenv("WANDB_ENTITY");
        project = System.getenv("WANDB_PROJECT");
        projectId = entity + "/" + project;
        authHeader = "Basic " + Base64.getEncoder().encodeToString(
                ("api:" + System.getenv("WANDB_API_KEY")).getBytes(StandardCharsets.UTF_8));

        String datasetName = "recipe-10-dataset-java-" + Instant.now().getEpochSecond();
        String datasetDescription = "Capital cities for evaluation (run at " + Instant.now() + ")";
        // Rows use LinkedHashMap (not Map.of) on purpose: Map.of has no defined
        // iteration order, so Jackson could emit answer before question — a
        // different byte-shape than the other ports. Since the rows Table is
        // content-addressed by row bytes, a divergent shape pollutes the shared
        // Table with duplicate rows. LinkedHashMap preserves question-then-answer.
        List<Map<String, Object>> datasetRows = List.of(
                qaRow("What is the capital of France?", "Paris"),
                qaRow("What is the capital of Spain?", "Madrid"),
                qaRow("What is the capital of Italy?", "Rome"));

        // Create the Dataset. v2 path; entity + project go into the URL.
        JsonNode created = post("/v2/" + entity + "/" + project + "/datasets", Map.of(
                "name", datasetName,
                "description", datasetDescription,
                "rows", datasetRows));
        String objectId = created.get("object_id").asText();
        String digest = created.get("digest").asText();
        System.out.println("Created: object_id=" + objectId + " digest=" + digest.substring(0, 12)
                + "… version=" + created.path("version_index").asText());

        // Read Dataset metadata back. GET, with object_id + digest in the URL.
        JsonNode dataset = get("/v2/" + entity + "/" + project + "/datasets/" + objectId + "/versions/" + digest);
        if (!datasetName.equals(dataset.path("name").asText()))
            fail("name: " + dataset.get("name"));
        if (!datasetDescription.equals(dataset.path("description").asText()))
            fail("description: " + dataset.get("description"));
        if (!objectId.equals(dataset.path("object_id").asText()))
            fail("object_id drift: " + dataset.get("object_id"));
        if (!digest.equals(dataset.path("digest").asText()))
            fail("digest drift: " + dataset.get("digest"));
        String rowsRef = dataset.path("rows").asText();
        System.out.println("Read:    name=\"" + dataset.path("name").asText() + "\" rows_ref=\"" + rowsRef + "\"");

        // The rows field is a reference to a Table (weave:///.../table/{digest}).
        // Parse the table digest so we can /table/query it; tolerate a bare digest.
        Matcher m = TABLE_REF.matcher(rowsRef);
        String tableDigest = m.find() ? m.group(1) : rowsRef;
        System.out.println("Table digest: " + tableDigest.substring(0, Math.min(12, tableDigest.length())) + "…");

        // Query the actual rows.
        JsonNode queried = post("/table/query", Map.of("project_id", projectId, "digest", tableDigest));
        JsonNode rows = queried.path("rows");

        // --- verification ---
        // Row count + per-row content (under `val`) must match what we wrote.
        if (rows.size() != datasetRows.size())
            fail("row count: " + rows.size() + " vs " + datasetRows.size());
        for (int i = 0; i < datasetRows.size(); i++) {
            if (!MAPPER.valueToTree(datasetRows.get(i)).equals(rows.get(i).get("val")))
                fail("row " + i + " val: " + rows.get(i).get("val") + " vs " + datasetRows.get(i));
        }
        System.out.println("Verified: " + rows.size() + " rows match (first: " + rows.get(0).get("val") + ")");
    }
}

File: golang/10_create_dataset.go
// Recipe 10: create a Dataset and read its rows back.
//
// Demonstrates the v2 Dataset endpoints plus the Table read needed to
// walk the rows:
//   POST /v2/{entity}/{project}/datasets
//       -> create the Dataset, returns (object_id, digest, version_index)
//   GET  /v2/{entity}/{project}/datasets/{object_id}/versions/{digest}
//       -> read Dataset metadata, including a *reference* to its rows
//   POST /table/query
//       -> read the actual rows out of the referenced Table
//
// Wire-level points worth knowing:
//
//   - These are v2 endpoints under /v2/{entity}/{project}/datasets, not a
//     v1-style /datasets/create. Entity + project live in the URL path.
//     Read uses GET (the rest of the API is POST-only); create uses POST.
//   - A Dataset is addressed by (object_id, digest) and is content-
//     addressed — identical (name, rows) collapses to the same version.
//     The name is stamped with a per-run Unix timestamp so every run
//     exercises the write path rather than resolving to an existing object.
//   - The read response's `rows` field is a *reference string* to the
//     underlying Table, not the row data. Parse the table digest out of
//     it and call /table/query. Rows come back wrapped as {digest, val,
//     original_index?} — the actual content lives under `val`.
//
// Run:
//   go run golang/10_create_dataset.go

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

var (
	entity    = os.Getenv("WANDB_ENTITY")
	project   = os.Getenv("WANDB_PROJECT")
	projectID = entity + "/" + project
	baseURL   = getenv("WEAVE_SERVICE_URL", "https://trace.wandb.ai")
	apiKey    = os.Getenv("WANDB_API_KEY")
	client    = &http.Client{}

	tableRefRe = regexp.MustCompile(`/table/([A-Za-z0-9_-]+)$`)
)

func getenv(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func fatal(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
	os.Exit(1)
}

// do issues a request with auth, optionally with a JSON body, and decodes
// the JSON response. Any non-2xx is fatal. post/get wrap it.
func do(method, path string, body map[string]any) map[string]any {
	var reader io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			fatal("encode %s: %v", path, err)
		}
		reader = bytes.NewReader(payload)
	}
	req, err := http.NewRequest(method, baseURL+path, reader)
	if err != nil {
		fatal("request %s: %v", path, err)
	}
	req.SetBasicAuth("api", apiKey)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	res, err := client.Do(req)
	if err != nil {
		fatal("%s %s: %v", method, path, err)
	}
	defer res.Body.Close()
	raw, _ := io.ReadAll(res.Body)
	if res.StatusCode/100 != 2 {
		fatal("HTTP %d for %s %s: %s", res.StatusCode, method, path, raw)
	}
	if len(raw) == 0 {
		return map[string]any{}
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		fatal("decode %s: %v", path, err)
	}
	return out
}

func post(path string, body map[string]any) map[string]any { return do(http.MethodPost, path, body) }
func get(path string) map[string]any                       { return do(http.MethodGet, path, nil) }

// qaRow is a dataset row. Struct fields (not a map) so JSON preserves
// question-before-answer key order, matching the other language ports.
type qaRow struct {
	Question string `json:"question"`
	Answer   string `json:"answer"`
}

func main() {
	var missing []string
	for _, k := range []string{"WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT"} {
		if os.Getenv(k) == "" {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		fatal("Missing required env vars: %s. See ../README.md#setup.", strings.Join(missing, ", "))
	}

	datasetName := fmt.Sprintf("recipe-10-dataset-golang-%d", time.Now().Unix())
	datasetDescription := fmt.Sprintf("Capital cities for evaluation (run at %s)", time.Now().UTC().Format(time.RFC3339Nano))
	// Rows are sent as a struct (not a map) on purpose: encoding/json emits
	// struct fields in declaration order, so each row serializes as
	// {"question": ..., "answer": ...}. A map would sort keys alphabetically
	// (answer before question), producing a different byte-shape than the
	// other ports — and since the rows Table is content-addressed by row
	// bytes, a divergent shape pollutes the shared Table with duplicate rows.
	datasetRows := []qaRow{
		{"What is the capital of France?", "Paris"},
		{"What is the capital of Spain?", "Madrid"},
		{"What is the capital of Italy?", "Rome"},
	}

	// Create the Dataset. v2 path; entity + project go into the URL.
	created := post(fmt.Sprintf("/v2/%s/%s/datasets", entity, project), map[string]any{
		"name":        datasetName,
		"description": datasetDescription,
		"rows":        datasetRows,
	})
	objectID := created["object_id"].(string)
	digest := created["digest"].(string)
	fmt.Printf("Created: object_id=%s digest=%s… version=%v\n", objectID, digest[:12], created["version_index"])

	// Read Dataset metadata back. GET, with object_id + digest in the URL.
	dataset := get(fmt.Sprintf("/v2/%s/%s/datasets/%s/versions/%s", entity, project, objectID, digest))
	if dataset["name"] != datasetName {
		fatal("name: %v", dataset["name"])
	}
	if dataset["description"] != datasetDescription {
		fatal("description: %v", dataset["description"])
	}
	if dataset["object_id"] != objectID {
		fatal("object_id drift: %v", dataset["object_id"])
	}
	if dataset["digest"] != digest {
		fatal("digest drift: %v", dataset["digest"])
	}
	rowsRef := dataset["rows"].(string)
	fmt.Printf("Read:    name=%q rows_ref=%q\n", dataset["name"], rowsRef)

	// The rows field is a reference to a Table (weave:///.../table/{digest}).
	// Parse the table digest so we can /table/query it; tolerate a bare digest.
	tableDigest := rowsRef
	if m := tableRefRe.FindStringSubmatch(rowsRef); m != nil {
		tableDigest = m[1]
	}
	fmt.Printf("Table digest: %s…\n", tableDigest[:min(12, len(tableDigest))])

	// Query the actual rows.
	queried := post("/table/query", map[string]any{"project_id": projectID, "digest": tableDigest})
	rows, _ := queried["rows"].([]any)

	// --- verification ---
	// Row count + per-row content (under `val`) must match what we wrote.
	if len(rows) != len(datasetRows) {
		fatal("row count: %d vs %d", len(rows), len(datasetRows))
	}
	for i, expected := range datasetRows {
		row, _ := rows[i].(map[string]any)
		val, _ := row["val"].(map[string]any)
		if val["question"] != expected.Question || val["answer"] != expected.Answer {
			fatal("row %d val: %v vs %+v", i, row["val"], expected)
		}
	}
	first, _ := rows[0].(map[string]any)
	fmt.Printf("Verified: %d rows match (first: %v)\n", len(rows), first["val"])
}
